File: design-pattern/facade-pattern/facade.ts
class PowerSupply {
    on(): void {
        console.log("Power Supply is ON");
    }
};


class Memory {
    checkMemory(): void {
        console.log("Memory is available");
    }
};

class CPU {
    cpuFan(): void {
        console.log("CPU fan is running");
    }
};

class HardDisk {
    loadOS(): void {
        console.log("Loading Operating System from Hard Disk");
    }
};

class BIOS {
    initailze(cpu: CPU, memory: Memory): void {
        cpu.cpuFan();
        memory.checkMemory();
    }
};

class Facade {
    private cpu: CPU;
    private memory: Memory;
    private os: HardDisk;
    private bios: BIOS;
    private ps: PowerSupply;

    constructor() {
        this.ps = new PowerSupply();
        this.cpu = new CPU();
        this.memory = new Memory();
        this.os = new HardDisk();
        this.bios = new BIOS();
    };

    startComputer(): void {
        this.ps.on();
        this.bios.initailze(this.cpu, this.memory);
        this.os.loadOS();
    }
};

const facade = new Facade();
facade.startComputer();

File: design-pattern/facade-pattern/facade.cpp
#include <iostream>

using namespace std;

class PowerSupply
{
public:
    void on()
    {
        cout << "Power Supply On" << endl;
    }
};

class CPU
{
public:
    void start()
    {
        cout << "CPU Started" << endl;
    }
};

class Memory
{
public:
    void load()
    {
        cout << "Memory Loaded" << endl;
    }
};

class HardDrive
{
public:
    void read()
    {
        cout << "Hard Drive Read" << endl;
    }
};

class Bios
{
public:
    void initialize(Memory &memory, CPU &cpu, HardDrive &hardDrive)
    {
        cout << "Bios Initialized" << endl;
        memory.load();
        cpu.start();
        hardDrive.read();
    }
};

class Facade
{
private:
    PowerSupply powerSupply;
    CPU cpu;
    Memory memory;
    HardDrive hardDrive;

public:
    void startComputer()
    {
        powerSupply.on();
        Bios bios;
        bios.initialize(memory, cpu, hardDrive);
    }
};

int main(void)
{
    Facade facade;
    facade.startComputer();
    return 0;
}